File: Data Structure/C++/BSTree.cpp
#include <iostream>
template <typename K, typename V>
class binarySearchTree
{
    private:
    struct node;
    void clear_(node*);
    node* root;
    size_t count;
    public:
    binarySearchTree( node* root_ = nullptr, const size_t& cnt = 0) : root(root_), count(cnt){}
    ~binarySearchTree();
    binarySearchTree(const binarySearchTree&);
    binarySearchTree& operator=(const binarySearchTree&);
    node* insert(const K&, const V&);
    void remove(const K&);
    V search(const K&);
    void inOrderTraversal(node*);
    void preOrderTraversal(node*);
    void postOrderTraversal(node*);
    node* getRoot();
    V getMax(node*) const;
    V getMin(node*) const;
};

template <typename K, typename V>
struct binarySearchTree<K, V>::node
{
    K key;
    V value;
    node* left, *right, *parent;
    binarySearchTree* host;
    node(const K& key_ = K(), const V& value_ = V(), node* left_ = nullptr, node* right_ = nullptr, node* parent_ = nullptr, binarySearchTree* bts = nullptr) :
    key(key_), value(value_), left(left_), right(right_), parent(parent_), host(bts){}
    ~node(){}
    node* insert_(const K&, const V&);
    void remove_(const K&);
    V search_(const K&);
    
};  

template <typename K, typename V>
typename  binarySearchTree<K, V>::node* binarySearchTree<K, V>::node::insert_(const K& nkey, const V& nval)
{
    if(nkey > key) 
    {
        if(right != nullptr) right -> insert_(nkey, nval);
        else return right = new node(nkey, nval, nullptr, nullptr, this, host);
    } else if(nkey < key)
    {
        if(left != nullptr) left -> insert_(nkey, nval);
        else return left = new node(nkey, nval, nullptr, nullptr, this, host);
    } else
    {
        key = nkey;
        value = nval;
        return this;
    }
    
}

template <typename K, typename V>
void binarySearchTree<K, V>::node::remove_(const K& nkey)
{
    if(nkey > key)
    {
        if(right != nullptr) right -> remove_(nkey);
        else throw std::string("No element");
    } else if(nkey < key)
    {
        if(left != nullptr) left -> remove_(nkey);
        else throw std::string("No element");
    } else
    {
        if(left == nullptr && right == nullptr)
        {
            if(parent == nullptr)
            {
                host -> root = nullptr;
            } else if(parent -> left && parent -> left -> key == nkey)
            {
                parent -> left = nullptr;
            } else if(parent -> right && parent -> right -> key == nkey)
            {
                parent -> right = nullptr;
            }       
        } else if(left != nullptr && right == nullptr)
        {
            if(parent == nullptr)
            {
                host -> root = left;
            } else if(parent -> left && parent -> left -> key == nkey)
            {
                parent -> left = left;
            } else if(parent -> right && parent -> right -> key == nkey)
            {
                parent -> right = left;
            }
            left -> parent = parent;
        } else if(left == nullptr && right != nullptr)
        {
            if(parent == nullptr)
            {
                host -> root = right;
            } else if(parent -> left && parent -> left -> key == nkey)
            {
                parent -> left = right;
            } else if(parent -> right && parent -> right -> key == nkey)
            {
                parent -> right = right;
            }
            right -> parent = parent;
        } else
        {
            if(right -> left == nullptr)
            {
                right -> parent = parent;
                right -> left = left;
                left -> parent = right;
                
                if(parent == nullptr) host -> root = right;
            } else 
            {
                node* lastLeft = right;
                while(lastLeft -> left)
                {
                    lastLeft = lastLeft -> left;
                }
                lastLeft -> parent -> left = nullptr;
                
                lastLeft -> left = left;
                lastLeft -> right = right;
                lastLeft -> parent = parent;
                
                if(parent == nullptr) host -> root = lastLeft;
            }
        }
        delete this;
    }
    
}

template <typename K, typename V>
V binarySearchTree<K, V>::node::search_(const K& nkey)
{
    if(nkey > key)
    {
        if (right != nullptr) return right -> search_(nkey);
        else throw std::string("No element");
    } else if(nkey < key)
    {
        if (left != nullptr) return left -> search_(nkey);
        else throw std::string("No element");
    } else
    {
        return value;
    }
}

template <typename K, typename V>
binarySearchTree<K, V>::~binarySearchTree()
{
    if(root != nullptr)
    {
      clear_(root);  
    }
    root = nullptr;
    count = 0;
}

template <typename K, typename V>
void binarySearchTree<K, V>::clear_(node* nd)
{
    if(nd != nullptr)
    {
        if(nd -> left != nullptr) clear_(nd -> left);
        if(nd -> right != nullptr) clear_(nd -> right);
        delete nd;
    }
}

template <typename K, typename V>
typename binarySearchTree<K, V>::node* binarySearchTree<K, V>::insert(const K& nkey, const V& nval)
{
    if(root == nullptr)
    {
        root = new node(nkey, nval, nullptr, nullptr, nullptr, this);
        return root;
    }
    return root -> insert_(nkey, nval);
}

template <typename K, typename V>
void binarySearchTree<K, V>::remove(const K& nkey)
{
    if(root == nullptr)
    {
        return;
    }
    root -> remove_(nkey);
}

template <typename K, typename V>
V binarySearchTree<K, V>::search(const K& nkey)
{
    if(root == nullptr)
    {
        throw std::string("No element");
    }
    return root -> search_(nkey);
}


template <typename K, typename V>
void binarySearchTree<K, V>::inOrderTraversal(binarySearchTree<K, V>::node* nd)
{
    if (nd != nullptr)
    {
        inOrderTraversal(nd -> left);
         std::cout<<nd -> value<<" ";
        inOrderTraversal(nd -> right);
    }
    std::cout<<std::endl;
}

template <typename K, typename V>
void binarySearchTree<K, V>::preOrderTraversal(binarySearchTree<K, V>::node* nd)
{
    
    if (nd != nullptr)
    {
        std::cout<<nd -> value<<" ";
        inOrderTraversal(nd -> left);
        inOrderTraversal(nd -> right);
    }
    std::cout<<std::endl;
}

template <typename K, typename V>
void binarySearchTree<K, V>::postOrderTraversal(node* nd)
{
      if (nd != nullptr)
    {
          
        inOrderTraversal(nd -> left);
        inOrderTraversal(nd -> right);
        std::cout<<nd -> value<<" ";
    }
    std::cout<<std::endl;
}

template <typename K, typename V>
typename binarySearchTree<K, V>::node* binarySearchTree<K, V>::getRoot()
{
   return root;
}
   

template <typename K, typename V>
V binarySearchTree<K, V>::getMin(node *nd) const
{
    if (nd -> left == nullptr)
    {
        return nd -> value;
    }
    return getMin(nd -> left);
}

template <typename K, typename V>
V binarySearchTree<K, V>::getMax(node *nd) const
{
    if (nd -> right == nullptr)
    {
        return nd -> value;
    }
    return getMax(nd -> right);
}

int main()
{
}

File: Data Structure/Python/BSTree.py
class BSTree():
    class Node():
        def __init__(self, key=0, val=0, left = None, right = None, parent = None):
            self.key = key
            self.val = val
            self.left = left
            self.right = right
            self.parent = parent
            
        def isRoot(self):
            return not self.parent
        
        def idLeaf(self):
            return not (self.right or self.left)
        
    def __init__(self):
        self.root = None
        self.size = 0
        
    def getRoot(self):
        return self.root    
        
    def length(self):
        return self.size
    
    def add(self, key, val):
        if self.root:
            self._add(key, val, self.root)
        else:
            self.root = self.Node(key, val)
        self.size += 1
        
    def _add(self, key, val, node):
        if key < node.key:
            if node.left != None:
                self._add(key, val, node.left)
            else:
                node.left = self.Node(key, val, parent = node)
        else:
            if node.right != None:
                self._add(key, val, node.right)
            else:
                node.right = self.Node(key, val, parent = node)
                
    def search(self, key):
        if self.root:
            element = self._search(key, self.root)
            if element != None:
                return element.val
            else:
                return None
        else:
            return None
        
    def _search(self, key, node):
        if not node:
            return None
        elif node.key == key:
            return node
        elif key < node.key:
            return self._search(key, node.left)
        else:
            return self._search(key, node.right)
        
    def remove(self, key):
        if self.size > 1:
            nodeToRemove = self._search(key, self.root)
            if nodeToRemove:
                self._remove(nodeToRemove)
                self.size -= 1
            else:
                raise KeyError('Invalid key')
        elif self.size == 1 and self.root.key == key:
            self.root = None
            self.size -= 1
        else:
            raise KeyError('Invalid key')
      
    def _remove(self, node):
            if node.left == None and node.right == None:
                if node.parent == None:
                    self.root = None
                elif node.parent.right and node.parent.right.key == node.key:
                    node.parent.right = None
                elif node.parent.left and node.parent.left.key == node.key:
                    node.parent.left = None
            elif node.left != None and node.right == None:
                if node.parent == None:
                    self.root = node.left
                elif node.parent.left and node.parent.left.key == node.key:
                    node.parent.left = node.left
                elif node.parent.right and node.parent.right.key == node.key:
                    node.parent.right = node.left 
                node.left.parent = node.parent
         
            elif node.left == None and node.right != None:
                if node.parent == None:
                    self.root = right
                elif node.parent.left and node.parent.left.key == node.key:
                    node.parent.left = node.right
                elif node.parent.right and node.parent.right.key == node.key:
                    node.parent.right = node.right
                node.right.parent = node.parent
            else: 
                if node.right.left == None:
                    node.right.parent = node.parent
                    node.right.left = node.left
                    node.left.parent = node.right
                    if node.parent == None: self.root = node.right
                else:
                    lastLeft = node.right
                    while lastLeft.left:
                        lastLeft = lastLeft.left
                    lastLeft.parent.left = None
                    lastLeft.left = node.left
                    lastLeft.right = node.right
                    lastLeft.parent = node.parent
                    if node.parent == None: self.root = lastLeft
              
    def inOrderTraversal(self, node):
        if node != None:
           self.inOrderTraversal(node.left)
           print(node.val)
           self.inOrderTraversal(node.right)
        print()
        
    def preOrderTraversal(self, node):
        if node != None:
           print(node.val) 
           self.inOrderTraversal(node.left)
           self.inOrderTraversal(node.right)
        print()
    
    def postOrderTraversal(self, node):
            if node != None:
               self.inOrderTraversal(node.left)
               self.inOrderTraversal(node.right)
               print(node.val) 
            print()
    
    def findMin(self):
        curr = self.root
        while curr.left:
            curr = curr.left
        return curr
